bin/getCryptoData.js

```diff
@@ -83,76 +92,88 @@ function getIcon(coin) {
-        !furtherResults.find(coin => coin.symbol === c.symbol)
-      ) {
-        newList.push(c);
-      }
-    });
-    totalCoins = newList.length;
-    results.data = newList;
-
-    results.data.sort((a, b) => {
-      if (a.symbol < b.symbol) {
-        return -1;
-      }
-      if (a.symbol > b.symbol) {
-        return 1;
-      }
-
-      return 0;
-    });
-
-    const callGetIcon = (index = 0) => {
-      getIcon(results.data[index]);
-      if (index + 1 < results.data.length) {
-        // Give some space between requests, otherwise kicking off so many at once
-        // leads to many of them failing.
-        setTimeout(() => callGetIcon(index + 1), 100);
-      }
-    };
-    callGetIcon();
-
-    if (results.data.length) {
-      console.log('Updating the en_US.json translation file...');
-      fs.readFile(translationFile, (err, data) => {
-        if (err) {
-          logError(`There was an error processing the translation file: ${err}`);
-          return;
-        }
-
-        const parsed = JSON.parse(data);
-        results.data.forEach(c => {
-          if (whitelist[c.symbol] && whitelist[c.symbol] !== c.id) return;
-          parsed.cryptoCurrencies[c.symbol] = c.name;
-        });
-        const jsonConfig = {
-          type: 'space',
-          size: 2,
-        };
-        fs.writeFile(translationFile, jsonformat(parsed, jsonConfig), writeErr => {
-          if (err) {
-            logError(`There was an error writing the translation file: ${writeErr}`);
-            return;
+console.log('Obtaining whitelist...');
+
+getWhitelist()
+  .then(
+    wl => {
+      whitelist = wl;
+      console.log('Whitelist obtained...');
+      draft('Obtaining crypto icons...');
+
+      getCoinList()
+        .then(results => {
+          // remove duplicates from the data set using our whitelist
+          const newList = [];
+          results.data.forEach((c, i) => {
+            const furtherResults = results.data.slice(i + 1);
+            if (
+              !(whitelist[c.symbol] && whitelist[c.symbol] !== c.id) &&
+              !furtherResults.find(coin => coin.symbol === c.symbol)
+            ) {
+              newList.push(c);
+            }
+          });
+          totalCoins = newList.length;
+          results.data = newList;
+
+          results.data.sort((a, b) => {
+            if (a.symbol < b.symbol) {
+              return -1;
+            }
+            if (a.symbol > b.symbol) {
+              return 1;
+            }
+
+            return 0;
+          });
+
+          const callGetIcon = (index = 0) => {
+            getIcon(results.data[index]);
+            if (index + 1 < results.data.length) {
+              // Give some space between requests, otherwise kicking off so many at once
+              // leads to many of them failing.
+              setTimeout(() => callGetIcon(index + 1), 50);
+            }
+          };
+          callGetIcon();
```

Comment:
Will this fire if results.data turns out to be empty, and attempt to call getIcon on undefined? Should it have a if (results.data.length) check?